source/_posts/2017-08-16-Getting-Stack-Traces-in-Swift-on-Linux.md

```diff
@@ -0,0 +1,125 @@
+
+Foundation in Swift provides a convenient `Thread.callStackReturnAddresses()`, but, like most things, this API is not available on Linux. To get around this, we have to write the implementation in C and then expose the implementation to Swift using a [modulemap](https://clang.llvm.org/docs/Modules.html). Doing so is not that complicated, but we ended up hitting a few roadblocks on the Linux side.
+
+We started out the implementation by taking advantage of [backtrace(3)](http://man7.org/linux/man-pages/man3/backtrace.3.html) that's available to *nix systems. Backtrace is a convenient function that just takes a buffer and its size and spits out a an array of pointers that is `<= BUFF_LEN` in length. There is even an equally convenient function `backtrace_symbols ` that can take these array of pointers and translate them into strings. The downside of `backtrace_symbols` is that its output is not defined by the standard. So, the resulting stack traces aren't the same across all implementations. On top of that, the output contains a lot of extra noise, such as address and offset information.
+
+## Getting a stack trace
+
+Our implementation of `backtrace_symbols` is relatively straightforward. We just map over the buffer using [dladdr(3)](http://man7.org/linux/man-pages/man3/dladdr.3.html) to get information about the address. If there is no metadata for the given symbol, the name defaults to `???`. The implementation could be more performant by iterating over the collection twice and using the first pass to calculate how much memory we need and then allocate a single buffer big enough to fit all of this information. Then iterate over the buffer again `memcpy`ing the strings into our pool. We did not use this approach because extracting strings from a single, pool-allocated, buffer is much more involved than extracting them from a nested pointer.
+
+```
+char ** get_symbols_for_backtrace(void * const *buffer, int size) {
+    int i;
+    char ** result;
+    
+    result = malloc(sizeof(char *) * size);
+    if (result == NULL) return NULL;
+    
+    for (i = 0; i <= size; ++i) {
+        Dl_info info;
+        dladdr(buffer[i], &info);
+        
+        char *name = "???";
+        if (info.dli_sname) {
+            name = (char *)info.dli_sname;
+        }
+
+        result[i] = strdup(name);
+    }
+    
+    return result;
+}
+```
+
+On Linux the symbol `dladdr` is a non-standard extension; therefore, it is only exposed by the feature test macro `_GNU_SOURCE`. Sadly, Swift imports glibc before you have the opportunity to define it ([SR-328](https://bugs.swift.org/browse/SR-328)). This can be worked around by invoking the Swift compiler with `-Xcc -D_GNU_SOURCE`. We ran into some issues when trying defining this macro on some versions of Linux and decide to build the stack trace library independent of the Swift package and then dynamically link it.
```

Comment:
You're the English expert, but I'm just checking that it's not a typo - ".. of Linux and decide to build .." decide or decided?